llvm/lib/Target/AMDGPU/AMDGPULoopAlign.cpp

```diff
@@ -0,0 +1,153 @@
+//===----- AMDGPULoopAlign.cpp - Generate loop alignment directives  -----===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// Inspect a basic block and if certain conditions are met then align to 32
+// bytes.
+//===----------------------------------------------------------------------===//
+
+#include "AMDGPULoopAlign.h"
+#include "AMDGPU.h"
+#include "AMDGPUTargetMachine.h"
+#include "GCNSubtarget.h"
+#include "SIMachineFunctionInfo.h"
+#include "llvm/CodeGen/MachineLoopInfo.h"
+using namespace llvm;
+
+#define DEBUG_TYPE "amdgpu-loop-align"
+
+static cl::opt<bool>
+    DisableLoopAlign("disable-amdgpu-loop-align", cl::init(false), cl::Hidden,
+                     cl::desc("Disable AMDGPU loop alignment pass"));
+
+namespace {
+
+class AMDGPULoopAlign {
+private:
+  MachineLoopInfo &MLI;
+
+public:
+  AMDGPULoopAlign(MachineLoopInfo &MLI) : MLI(MLI) {}
+
+  struct BasicBlockInfo {
+    // Offset - Distance from the beginning of the function to the beginning
+    // of this basic block.
+    uint64_t Offset = 0;
+    // Size - Size of the basic block in bytes
+    uint64_t Size = 0;
+  };
+
+  void generateBlockInfo(MachineFunction &MF,
```

Comment:
I really don't want to add yet another instance of a pass doing its own code size computations. For the purpose of this pass you should be able to just set the alignment on the block. getInstSIzeInBytes isn't completely accurate either